packages/api/src/get-repo-orgs.js

```diff
@@ -0,0 +1,97 @@
+
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+
+// class to receive POSTS with build information
+
+class GetRepoOrgsHandler {
+  constructor (app, client) {
+    this.app = app;
+    this.client = client;
+  }
+
+  // ex: bar -> bas
+  alphabeticallyIncrement (inputStr) {
+    const strFrontCode = inputStr.slice(0, inputStr.length - 1);
+    const strEndCode = inputStr.slice(inputStr.length - 1, inputStr.length);
+    return strFrontCode + String.fromCharCode(strEndCode.charCodeAt(0) + 1);
+  }
+
+  removeDuplicates (results) {
+    const usedIds = new Set();
+    const newResults = [];
+    results.forEach(result => {
+      if (!usedIds.has(result.repoId)) {
+        usedIds.add(result.repoId);
+        newResults.push(result);
+      }
+    });
+    return newResults;
+  }
+
+  // TODO order by flakyness
+  listen () {
+    // return the different environments for a build - all parameters and possible values for next query
+    this.app.get('/api/reposearch', async (req, res, next) => {
+      try {
+        const limit = parseInt(req.query.limit || 10);
+        if (!req.query.startswith) { // case with no query
+          const snapshot = await this.client.collection(global.headCollection).limit(limit).get();
+          res.send(snapshot.docs.map(doc => doc.data()));
+        } else {
+          // get 10 results starting with name or org/name
+          const results = [];
+
+          const snapshotName = await this.client.collection(global.headCollection)
+            .where('name', '>=', req.query.startswith)
+            .where('name', '<', this.alphabeticallyIncrement(req.query.startswith))
+            .limit(limit).get();
+          snapshotName.docs.forEach(doc => results.push(doc.data()));
+
+          const snapshotOrg = await this.client.collection(global.headCollection)
+            .where('repoId', '>=', req.query.startswith)
+            .where('repoId', '<', this.alphabeticallyIncrement(req.query.startswith))
+            .limit(limit).get();
+          snapshotOrg.docs.forEach(doc => results.push(doc.data()));
+
+          res.send(this.removeDuplicates(results));
```

Comment:
I think we might be able to simplify this if we assume that we know the organization, so are just querying for the repos under that organization. I think we could assume that on the front end, someone has indicated the org they're interested in before hitting this route.

Reply:
I was imagining initially it would be a repositorycentric view, so you would want to search for repositories from the landing page by typing strings like "Googlecloudplatform/fl" or "flaky-se" or "Goog"